### PR TITLE
PLAT-191: Reset password

### PR DESCRIPTION
**Pull request by docktorrr**
## Purpose
We need to add reset password functionality to CoreUser

## Approach
- Endpoint for sending password reset request
- Endpoint for password reset confirmation and setting new password (we can do it one step like it is in Django built-in forms instead of separate token validation endpoint)
~~The proposal is to break the task into smaller PR's. In this PR the endpoint for sending password reset request is added.~~
~~The built-in Django mechanism is used for password resetting, for this purpose the built-in PasswordResetForm is overridden, to change sending email process (Messaging Service will be used).~~

### Open Questions and Pre-Merge TODOs
- [ ] Do we need separate token validation endpoint? We have to check token anyway while sending new passwords, so double token checking doesn't make sense for me.

### CODE REVIEW

File: workflow/views.py [hunk 1]
```diff
@@ -385,29 +385,48 @@ def perform_invite(self, serializer):
...
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(
+            {
+                'detail': 'The reset password link was sent successfully.',
+            },
+            status=status.HTTP_200_OK)
 
+    def send_email(self, email_address: str, subject: str, text_content: str, html_content: str = None) -> None:
```
Comment (jefmoura) [inline workflow/views.py]: Shouldn't we put this function out of the view set, so we can use it in other places?

File: workflow/views.py [hunk 2]
```diff
@@ -385,29 +385,48 @@ def perform_invite(self, serializer):
...
+        # TODO: make sending with messaging service when it's ready instead of direct e-mailing
```
Comment (jefmoura) [inline workflow/views.py]: I don't think we're gonna create a dependency between Bifrost and messaging service

File: workflow/views.py [hunk 3]
```diff
@@ -385,29 +385,48 @@ def perform_invite(self, serializer):
...
         msg = EmailMultiAlternatives(
-            subject='Application Access',  # TODO we need to make this dynamic
+            subject=subject,
             body=text_content,
             to=[email_address],
         )
-        msg.attach_alternative(html_content, "text/html")
+        if html_content:
+            msg.attach_alternative(html_content, "text/html")
```
Comment (jefmoura) [inline workflow/views.py]: Maybe you can return the response from `msg.send()`, so you know how many emails were sent and use it in validations and/or tests

File: workflow/serializers.py [hunk 4]
```diff
@@ -129,6 +130,31 @@ class CoreUserInvitationSerializer(serializers.Serializer):
...
+    email = serializers.EmailField()
+
+    def validate_email(self, value):
+        self._reset_form = CoreUserPasswordResetForm(data=self.initial_data)
+        if not self._reset_form.is_valid():
+            raise serializers.ValidationError(self._reset_form.errors)
+        return value
+
+    def save(self):
+        assert hasattr(self, '_reset_form'), "You should validate serializer before saving"
```
Comment (jefmoura) [inline workflow/serializers.py]: I think it's better to raise an error different of `AssertionError`

**Comment by docktorrr** [comment]
@jefmoura Now you can review. I've rewrote it without using built-in django form, because it needs to override a lot, so it doesn't make sense in my opinion.

File: workflow/serializers.py [hunk 5]
```diff
@@ -129,6 +136,69 @@ class CoreUserInvitationSerializer(serializers.Serializer):
...
 
 
+class CoreUserResetPasswordSerializer(serializers.Serializer):
+    """
+    Serializer for reset password request data
+    """
+    email = serializers.EmailField()
+
+    def save(self, **kwargs):
+        resetpass_url = urljoin(settings.FRONTEND_URL, settings.REGISTRATION_URL_PATH)
```
Comment (jefmoura) [inline workflow/serializers.py]: Why are you using `REGISTRATION_URL_PATH`?

File: workflow/serializers.py [hunk 6]
```diff
@@ -129,6 +136,69 @@ class CoreUserInvitationSerializer(serializers.Serializer):
...
+
+        return count
+
+
+class CoreUserResetPasswordConfirmSerializer(serializers.Serializer):
+    """
+    Serializer for reset password data
+    """
+    new_password1 = serializers.CharField(max_length=128)
+    new_password2 = serializers.CharField(max_length=128)
```
Comment (jefmoura) [inline workflow/serializers.py:172]: Should we use `style={'input_type': 'password'}` for these fields?

File: workflow/tests/test_coreuser.py [hunk 7]
```diff
@@ -230,3 +205,84 @@ def test_invitation_check(request_factory, org):
     assert response.data['email'] == TEST_USER_DATA['email']
     assert response.data['organization']['organization_uuid'] == \
         str(org.organization_uuid)
+
+
+@pytest.mark.django_db()
+def test_reset_password(request_factory, org_member):
```
Comment (jefmoura) [inline workflow/tests/test_coreuser.py]: Why don't you group the reset password tests in a class? E.g, https://github.com/Humanitec/bifrost/blob/master/web/tests/test_views.py#L55

File: workflow/tests/test_coreuser.py [hunk 8]
```diff
@@ -230,3 +205,84 @@ def test_invitation_check(request_factory, org):
...
+        'new_password2': test_password2,
+        'uid': uid,
+        'token': token,
+    }
+    request = request_factory.post(reverse('coreuser-reset-password-confirm'), data)
+    response = CoreUserViewSet.as_view({'post': 'reset_password_confirm'})(request)
+    assert response.status_code == 400  # validation error (password fields didn't match)
+
+
+from unittest import mock
```
Comment (jefmoura) [inline workflow/tests/test_coreuser.py]: Please, always import in the top of the file

**Review by jefmoura — DISMISSED** [review]
- It's not working for me. I tried to reset my password and when I access reset password confirm endpoint, it returns `Page Not Found`. I don't think we should ask the FE to always implement a way to get this value `Mw-546-28381e5f6a42b040e010` and break it into `uid` and `token`. We should have an endpoint that they can send it and we're gonna validate it as Django does.
firefox_screenshot_2019-02-26t10-13-11 782z

- The swagger schema doesn't match with the data we should send to the endpoints. We need to update the schema as well.
firefox_screenshot_2019-02-26t10-06-26 030z

File: workflow/tests/test_coreuser.py (hunk 8 above)
Comment (docktorrr) [reply workflow/tests/test_coreuser.py]: Oh, it was in debug purpose, forget to move it to the top... Strange that flake8 let it pass

File: workflow/serializers.py (hunk 6 above)
Comment (docktorrr) [reply workflow/serializers.py:172]: Why? I think no, the fields will be rendered by FE.

File: workflow/serializers.py (hunk 5 above)
Comment (docktorrr) [reply workflow/serializers.py]: Will fix it, thanks, in the test as well.

**Comment by docktorrr** [comment]
> * It's not working for me. I tried to reset my password and when I access reset password confirm endpoint, it returns `Page Not Found`. I don't think we should ask the FE to always implement a way to get this value `Mw-546-28381e5f6a42b040e010` and break it into `uid` and `token`....
  @jefmoura @glind  
- But this URL is incorrect, the token should be inside the post request. That's why 404
- `uid` and `token` combined - I did it just like Django do.
- FE just gets the token from the URL and passed it to BE in the confirm request as it is. They don't need to encode/decode/break anything.

**Comment by jefmoura** [comment]
@docktorrr These are the URLs the Django generates for resetting a password. Check the one with the red square.
firefox_screenshot_2019-02-26t11-01-19 184z

**Comment by docktorrr** [comment]
@jefmoura Yes, you are right, it seems that I came across some older version URL example. I fixed it to "/"

**Review by glind — APPROVED** [review]

**Pull request merged**